### PR TITLE
[CoroutineAccessors] Initial framing.

### PR DESCRIPTION
**Pull request by nate-chandler**
Initial steps towards `read` and `modify`.  Add a new feature flag.  Behind it, allow parsing the keywords.  Add mangling for these symbols.  For now, they are emitted just like `_read` and `_modify`.

### CODE REVIEW

**Comment by nate-chandler** [comment]
https://github.com/swiftlang/swift-syntax/pull/2853

@swift-ci please test

**Comment by nate-chandler** [comment]
https://github.com/swiftlang/swift-syntax/pull/2853

@swift-ci please test

**Comment by nate-chandler** [comment]
https://github.com/swiftlang/swift-syntax/pull/2855

@swift-ci please test

**Comment by nate-chandler** [comment]
https://github.com/swiftlang/swift-syntax/pull/2855

@swift-ci please test

**Comment by nate-chandler** [comment]
https://github.com/swiftlang/swift-syntax/pull/2855

@swift-ci please test

**Comment by nate-chandler** [comment]
https://github.com/swiftlang/swift-syntax/pull/2855

@swift-ci please test

**Comment by nate-chandler** [comment]
https://github.com/swiftlang/swift-syntax/pull/2855

@swift-ci please test linux platform

**Comment by nate-chandler** [comment]
https://github.com/swiftlang/swift-syntax/pull/2855

@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
https://github.com/swiftlang/swift-syntax/pull/2855

@swift-ci please test

**Comment by nate-chandler** [comment]
https://github.com/swiftlang/swift-syntax/pull/2855

@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
https://github.com/swiftlang/swift-syntax/pull/2855

@swift-ci please test

**Comment by nate-chandler** [comment]
https://github.com/swiftlang/swift-syntax/pull/2855

@swift-ci please test source compatibility

File: lib/Parse/ParseDecl.cpp [hunk 1]
```diff
@@ -7861,7 +7861,9 @@ static bool isAllowedWhenParsingLimitedSyntax(AccessorKind kind, bool forSIL) {
   case AccessorKind::WillSet:
   case AccessorKind::DidSet:
   case AccessorKind::Read:
+  case AccessorKind::Read2:
```
Comment (ahoppen) [inline lib/Parse/ParseDecl.cpp:7864]: Would it make sense to have more descriptive names here than `Read2`? For example, rename the exiting `Read` to `UnderscoreRead` and then use `Read` for `Read2`.

File: lib/Parse/ParseDecl.cpp [hunk 2]
```diff
@@ -7944,8 +7948,24 @@ static ParserStatus parseAccessorIntroducer(Parser &P,
...
+// call to a function named experimental_accessor passing a trailing closure.
+// To maintain compatibility, if the feature corresponding to
+// experimental_accessor is not enabled, do not parse the token as an
+// introducer, so that it can be parsed as such a function call.
+#define EXPERIMENTAL_COROUTINE_ACCESSOR(ID, KEYWORD, FEATURE)                  \
+  else if (P.Tok.getRawText() == #KEYWORD) {                                   \
+    if (!P.Context.LangOpts.hasFeature(Feature::FEATURE)) {                    \
+      if (featureUnavailable)                                                  \
+        *featureUnavailable = true;                                            \
+      if (isFirstAccessor) {                                                   \
```
Comment (ahoppen) [inline lib/Parse/ParseDecl.cpp:7962]: Shouldn’t this be the other way round? Ie. shouldn’t we allow these tokens on the first accessor but not on any following? Or am I missing something?

File: lib/Parse/ParseDecl.cpp (hunk 1 above)
Comment (nate-chandler) [reply lib/Parse/ParseDecl.cpp:7864]: I think Read2 and Modify2 are fine placeholders for now until the final names are decided on.

File: lib/Parse/ParseDecl.cpp (hunk 2 above)
Comment (nate-chandler) [reply lib/Parse/ParseDecl.cpp:7962]: No, we need to allow `read` and `modify` to be the names functions called with an implicit return from an implicit getter.  See `test/Parse/coroutine_accessors_ambiguity.swift`.

Comment (ahoppen) [reply lib/Parse/ParseDecl.cpp:7962]: Oh, do we backtrack if we return an error status from `parseAccessorIntroducer`?

Comment (nate-chandler) [reply lib/Parse/ParseDecl.cpp:7962]: Yeah, in `parseGetSet`.

Comment (ahoppen) [reply lib/Parse/ParseDecl.cpp:7962]: Great. Thanks for confirming.

**Pull request merged**